File: packages/java/src/main/java/io/flatwire/FlatChecked.java
package io.flatwire;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.function.Consumer;

/**
 * Partial-stream failure semantics for flatwire (checked streams), matching the
 * Python/Node reference. A streamed collection is wrapped in an envelope whose
 * terminal status is written LAST:
 *
 * <pre>
 *   {"items":[ e0, e1, ... ],"complete":true}
 *   {"items":[ e0, e1, ... ],"complete":false,"error":{"message":"...","type":"..."}}
 * </pre>
 *
 * so a consumer can tell clean completion, an in-band producer error after N
 * rows, and truncation apart. The wire is plain JSON, so a checked stream
 * written by any flatwire language decodes here and vice versa. See
 * docs/FAILURE.md.
 */
public final class FlatChecked {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private FlatChecked() {
    }

    /** The producer finished a checked stream with {@code complete:false}. */
    public static final class CheckedStreamException extends IOException {
        private final transient JsonNode error;

        public CheckedStreamException(String message, JsonNode error) {
            super(message);
            this.error = error;
        }

        /** The decoded error payload the producer wrote on the wire (may be null). */
        public JsonNode getError() {
            return error;
        }
    }

    /** The stream ended before a terminal status was written. */
    public static final class TruncatedStreamException extends IOException {
        public TruncatedStreamException(String message) {
            super(message);
        }
    }

    /**
     * Stream {@code items} inside a checked envelope, writing the terminal status
     * last. If serializing an element throws, a {@code complete:false} trailer
     * carrying the error is written before the original exception is re-thrown, so
     * the consumer can distinguish failure from truncation. Returns the element
     * count.
     */
    public static long encodeCheckedArray(Iterable<?> items, OutputStream out) throws IOException {
        out.write("{\"items\":[".getBytes(java.nio.charset.StandardCharsets.UTF_8));
        long count = 0;
        try {
            for (Object item : items) {
                if (count > 0) {
                    out.write(',');
                }
                out.write(MAPPER.writeValueAsBytes(item));
                count++;
            }
        } catch (RuntimeException | IOException ex) {
            java.util.Map<String, String> err = new java.util.LinkedHashMap<>();
            err.put("message", ex.getMessage() == null ? ex.toString() : ex.getMessage());
            err.put("type", ex.getClass().getSimpleName());
            out.write("],\"complete\":false,\"error\":".getBytes(java.nio.charset.StandardCharsets.UTF_8));
            out.write(MAPPER.writeValueAsBytes(err));
            out.write('}');
            throw ex;
        }
        out.write("],\"complete\":true}".getBytes(java.nio.charset.StandardCharsets.UTF_8));
        return count;
    }

    /**
     * Lazily read a checked envelope from {@code in}, handing each element to
     * {@code consumer} in turn (one at a time, so peak memory stays flat), then
     * enforce the terminal status. Backed by Jackson's streaming parser.
     *
     * @throws CheckedStreamException if the producer signalled {@code complete:false}
     * @throws TruncatedStreamException if the stream ended before a terminal status
     */
    public static <T> void decodeCheckedArray(InputStream in, Class<T> type, Consumer<T> consumer)
            throws IOException {
        try (JsonParser parser = MAPPER.getFactory().createParser(in)) {
            if (expect(parser) != JsonToken.START_OBJECT) {
                throw new IOException("flatwire: decodeCheckedArray expects a checked stream object");
            }
            if (expect(parser) != JsonToken.FIELD_NAME || !"items".equals(parser.currentName())) {
                throw new IOException("flatwire: checked stream must start with an \"items\" array");
            }
            if (expect(parser) != JsonToken.START_ARRAY) {
                throw new IOException("flatwire: checked stream \"items\" must be an array");
            }

            while (true) {
                JsonToken tok = next(parser);
                if (tok == null) {
                    throw new TruncatedStreamException("stream ended inside items array");
                }
                if (tok == JsonToken.END_ARRAY) {
                    break;
                }
                T element = MAPPER.readValue(parser, type);
                consumer.accept(element);
            }

            boolean seenComplete = false;
            boolean complete = false;
            JsonNode error = null;
            while (true) {
                JsonToken tok = next(parser);
                if (tok == null) {
                    throw new TruncatedStreamException("stream ended before terminal status");
                }
                if (tok == JsonToken.END_OBJECT) {
                    break;
                }
                if (tok == JsonToken.FIELD_NAME) {
                    String field = parser.currentName();
                    next(parser);
                    if ("complete".equals(field)) {
                        complete = parser.getBooleanValue();
                        seenComplete = true;
                    } else if ("error".equals(field)) {
                        error = MAPPER.readTree(parser);
                    } else {
                        parser.skipChildren();
                    }
                }
            }

            if (!seenComplete) {
                throw new TruncatedStreamException("stream ended before terminal status");
            }
            if (!complete) {
                String msg = error != null && error.has("message")
                        ? error.get("message").asText()
                        : "unknown stream error";
                throw new CheckedStreamException(msg, error);
            }
        }
    }

    private static JsonToken expect(JsonParser parser) throws IOException {
        JsonToken tok;
        try {
            tok = parser.nextToken();
        } catch (com.fasterxml.jackson.core.io.JsonEOFException eof) {
            throw new TruncatedStreamException("stream ended before items array");
        }
        if (tok == null) {
            throw new TruncatedStreamException("stream ended before items array");
        }
        return tok;
    }

    private static JsonToken next(JsonParser parser) throws IOException {
        try {
            return parser.nextToken();
        } catch (com.fasterxml.jackson.core.io.JsonEOFException eof) {
            return null;
        }
    }
}

File: packages/dotnet/FlatWire/FlatChecked.cs
using System.Text;
using System.Text.Json;

namespace FlatWire;

/// <summary>The producer finished a checked stream with <c>complete:false</c>.</summary>
public sealed class CheckedStreamException : Exception
{
    /// <summary>The decoded error payload the producer wrote on the wire.</summary>
    public JsonElement? Error { get; }

    public CheckedStreamException(string message, JsonElement? error = null) : base(message)
    {
        Error = error;
    }
}

/// <summary>The stream ended before a terminal status was written.</summary>
public sealed class TruncatedStreamException : Exception
{
    public TruncatedStreamException(string message) : base(message) { }
}

/// <summary>
/// Partial-stream failure semantics for flatwire (checked streams), matching the
/// Python/Node reference. A streamed collection is wrapped in an envelope whose
/// terminal status is written LAST:
/// <code>
///   {"items":[ e0, e1, ... ],"complete":true}
///   {"items":[ e0, e1, ... ],"complete":false,"error":{"message":"...","type":"..."}}
/// </code>
/// so a consumer can tell clean completion, an in-band producer error after N
/// rows, and truncation apart. The wire is plain JSON, so a checked stream written
/// by any flatwire language decodes here and vice versa. See docs/FAILURE.md.
/// </summary>
public static class FlatChecked
{
    private static readonly JsonSerializerOptions Options = new(JsonSerializerDefaults.Web);

    /// <summary>
    /// Stream <paramref name="items"/> inside a checked envelope, writing the
    /// terminal status last. If iterating the source throws, a
    /// <c>complete:false</c> trailer carrying the error is written before the
    /// original exception is re-thrown, so the consumer can distinguish failure
    /// from truncation. Returns the number of elements written.
    /// </summary>
    public static long EncodeCheckedArray<T>(IEnumerable<T> items, Stream destination)
    {
        destination.Write("{\"items\":["u8);
        long count = 0;
        try
        {
            foreach (var item in items)
            {
                if (count > 0) destination.Write(","u8);
                destination.Write(JsonSerializer.SerializeToUtf8Bytes(item, Options));
                count++;
            }
        }
        catch (Exception ex)
        {
            var err = new Dictionary<string, string>
            {
                ["message"] = ex.Message,
                ["type"] = ex.GetType().Name,
            };
            destination.Write("],\"complete\":false,\"error\":"u8);
            destination.Write(JsonSerializer.SerializeToUtf8Bytes(err, Options));
            destination.Write("}"u8);
            throw;
        }
        destination.Write("],\"complete\":true}"u8);
        return count;
    }

    /// <summary>
    /// Yield each element from a checked envelope, then enforce the terminal
    /// status. Throws <see cref="CheckedStreamException"/> if the producer
    /// signalled <c>complete:false</c>, and <see cref="TruncatedStreamException"/>
    /// if the stream ended before any terminal status. Peak memory stays bounded
    /// by the largest single element plus the small trailer.
    /// </summary>
    public static IEnumerable<T?> DecodeCheckedArray<T>(Stream source)
    {
        using var reader = new StreamReader(source, Encoding.UTF8, false, 65536, leaveOpen: true);
        var chunk = new char[65536];
        var s = new StringBuilder();
        var eof = false;

        bool More()
        {
            if (eof) return false;
            int n = reader.Read(chunk, 0, chunk.Length);
            if (n <= 0) { eof = true; return false; }
            s.Append(chunk, 0, n);
            return true;
        }

        // Consume the fixed header, then drop it so `s` holds only unconsumed
        // items/trailer bytes and stays bounded by the largest element.
        const string header = "{\"items\":[";
        while (s.Length < header.Length)
        {
            if (!More()) throw new TruncatedStreamException("stream ended before items array");
        }
        if (s.ToString(0, header.Length) != header)
            throw new InvalidDataException("DecodeCheckedArray: not a flatwire checked stream");
        s.Remove(0, header.Length);

        int pos = 0, depth = 0;
        bool inString = false, escape = false;

        while (true)
        {
            while (pos < s.Length)
            {
                char ch = s[pos];
                if (inString)
                {
                    if (escape) escape = false;
                    else if (ch == '\\') escape = true;
                    else if (ch == '"') inString = false;
                    pos++;
                    continue;
                }
                if (ch == '"') { inString = true; pos++; }
                else if (ch == '{' || ch == '[') { depth++; pos++; }
                else if (ch == ']' && depth == 0)
                {
                    var seg = s.ToString(0, pos).Trim();
                    if (seg.Length > 0) yield return JsonSerializer.Deserialize<T>(seg, Options);
                    pos++;
                    while (More()) { }
                    var trailer = s.ToString(pos, s.Length - pos).Trim();
                    Finish(trailer);
                    yield break;
                }
                else if (ch == '}' || ch == ']') { depth--; pos++; }
                else if (ch == ',' && depth == 0)
                {
                    var seg = s.ToString(0, pos).Trim();
                    if (seg.Length > 0) yield return JsonSerializer.Deserialize<T>(seg, Options);
                    s.Remove(0, pos + 1);
                    pos = 0;
                }
                else pos++;
            }
            if (!More()) throw new TruncatedStreamException("stream ended inside items array");
        }
    }

    private static void Finish(string trailer)
    {
        if (trailer.Length == 0)
            throw new TruncatedStreamException("stream ended before terminal status");
        using var doc = JsonDocument.Parse("{" + (trailer.StartsWith(',') ? trailer[1..] : trailer));
        var root = doc.RootElement;
        if (!root.TryGetProperty("complete", out var complete))
            throw new TruncatedStreamException("stream ended before terminal status");
        if (complete.ValueKind == JsonValueKind.False)
        {
            var err = root.TryGetProperty("error", out var e) ? e.Clone() : (JsonElement?)null;
            var msg = err is { ValueKind: JsonValueKind.Object } eo &&
                      eo.TryGetProperty("message", out var m) ? m.GetString() : "unknown stream error";
            throw new CheckedStreamException(msg ?? "unknown stream error", err);
        }
    }
}
